### PR TITLE
[WIP] Vertex cache

### PR DESCRIPTION
**Pull request by Pokechu22**
An implementation of the vertex cache, as described in patents [US6717577B1](https://patents.google.com/patent/US6717577B1/en) and [JP4644353B2](https://patents.google.com/patent/JP4644353B2/ja) (note that these are very early patents (1999), which probably don't match the real implementation).  Should fix [issue 11746](https://bugs.dolphin-emu.org/issues/11746).

-----

This is a very early draft.  It works for the penguins, but it's definitely not complete.

- [x] Supports caching
- [x] Supports invalidating (mostly)
- [ ] Implement for the JIT-based vertex loaders (currently I force the software vertex loader)
- [ ] Determine actual constants such as the cache size (0x20 or 0x80 bytes), number of lines, number of sets, memory latency, etc.
- [ ] Implement a proper least-recently-used (LRU) replacement algorithm/whatever is actually used.  Right now, if the cache is full (i.e. hasn't been invalidated explicitly in a while), it just replaces the first entry always.
- [ ] Handle memory latency, including the queue of pending reads (or at least fake it to reflect the way it affects which line is replaced)
- [ ] Handle stalls separately from misses (if that's actually a needed thing?)
- [ ] Handle perf query registers (CP commands 0x00/0x10/0x20, `VCACHE_METRIC_CHECK_L` registers, etc.)
- [ ] Handle the indexed load commands (which libogc says use the vertex cache)
- [ ] Make the code not terrible (e.g. make `VertexCache` a field of the vertex loader, as long as that doesn't conflict with the above)
- [ ] Actually hardware test stuff (which will probably require determining how the perf query registers work, and writing a non-empty replacement for `GX_SetVCacheMetric`)

It's likely that all of this code will eventually be thrown out and redone.

### CODE REVIEW

File: Source/Core/VideoCommon/VertexCache.cpp [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+namespace
+{
+u32 GetCacheLineOffset(u32 address)
+{
+  return (address >> CACHE_LINE_OFFSET_SHIFT) & CACHE_LINE_OFFSET_MASK;
+}
+
+u32 GetSetNumber(u32 address)
+{
+  return (address >> SET_NUMBER_SHIFT) & SET_NUMBER_SHIFT;
```
Comment (Warepire) [inline Source/Core/VideoCommon/VertexCache.cpp:87]: Should the latter SHIFT be MASK?

Comment (Pokechu22) [reply Source/Core/VideoCommon/VertexCache.cpp:87]: Yes, but that ends up not mattering here because both SHIFT and MASK are 7 in this case (but that does cause problems if the cache line size is changed).  (I actually caught this yesterday but hadn't pushed a fix yet).

**Comment by Pokechu22** [comment]
This morning I realized a rather terrific fact (in both the modern sense of the word as "very good" and the old sense of "causing terror"): the huge pile of [fifoci differences](https://fifo.ci/version/ca2b5be2cf85d3f7eaf56896737e6ac993dff233/) isn't an emulation issue, but in fact a (mostly) correct side effect of emulation of the vertex cache.  Specifically, due to how memory updates work, the fifoplayer may only record part of the used memory initially, and then more of it later on.  For the border in [mp4-vertexcache](https://fifo.ci/compare/7284561-7277936/), all parts of the border are drawn based on a position array at `009c3160`, but the fifo player updates that same array 5 times: first 0x48 bytes of it (0x009c3160-0x009c31a7), then 0x54 (0x009c3160-0x009c31b3), then 0x60 (0x009c3160-0x009c31bf), then 0xf0 (0x009c3160-0x009c324f) then 0x1b0 (0x009c3160-0x009c330f).  This isn't because the data changed, but instead because later primitives used higher indices, so it needed to record more data.  But the vertex cache only applies the first time, meaning an incomplete copy is cached, and that incomplete cached one is used instead of the later memory updates applying.  This is a fifoplayer-specific issue and not something that would apply to the regular game (the regular game almost certainly had already written the full range of memory before it started rendering).

A very similar thing happens on real hardware with the hardware fifoplayer on the first frame. In both Dolphin (using the HW fifoplayer and using Dolphin's fifoplayer) and on real hardware, things generally render correctly on subsequent loops of the first frame.  However, on real hardware, the hardware fifoplayer's lack of clearing memory results in additional garbage data; I've added a second set of images where all memory is `memset` to 0 first (there are other complications involving the way the hardware fifoplayer manages memory, since it can't use the same addresses games use since it has to care about its own code and data not being overwritten).  Also, the texture cache is affected by this as well as the vertex cache, resulting in other glitchy results.  All dolphin images are with this PR.

<table>
<tr><th>Frame</th><th>Memset</th><th width="40%">Hardware</th><th width="40%">Dolphin</th></tr>
<tr><th>1</th><th>No</th><td></td><td></td></tr>
<tr><th>1</th><th>Yes</th><td></td><td></td></tr>
<tr><th>2</th><th>No</th><td></td><td></td></tr>
<tr><th>2</th><th>Yes</th><td></td><td></td></tr>
</table>

In short: the way that fifologs handle memory updates causes issues with a proper implementation of the vertex cache.  These issues shouldn't affect actual games, but it does cause problems for FifoCI.  I don't think there's a good way of fixing these issues (they'll still happen with a better implementation of a vertex cache; we'd need to rework fifologs to fix them, and that would also involve replacing/fixing all of the existing fifologs).

**Comment by Pokechu22** [comment]
I've changed the cache to use 0x20 bytes per cache line, 32 cache lines per set, and 8 sets (still). 0x20 bytes is what libogc says, and it seems to be correct; the change from 8 to 32 cache lines is to try and use the same total memory but that's just a guess.  ([FifoCI](https://fifo.ci/version/51655b6bc033f8d2a094fa64200772568df7431f/)).

This has resulted in a much closer match for the weird behavior at the start of fifologs:

<table>
<tr><th>Frame</th><th>Memset</th><th width="40%">Hardware</th><th width="40%">Dolphin</th></tr>
<tr><th>1</th><th>No</th><td></td><td></td></tr>
<tr><th>1</th><th>Yes</th><td></td><td></td></tr>
<tr><th>2</th><th>No</th><td></td><td></td></tr>
<tr><th>2</th><th>Yes</th><td></td><td></td></tr>
</table>

I'm pretty sure all vertex coordinates match there (at least for the memset case); it's just the texture cache that's behaving differently now.

**Comment by Pokechu22** [comment]
I've rebased this PR atop #10384 to get a better idea of what differences are caused by the vertex cache versus the jank thing I've tried to implement to resolve the memory updates issue.  Here's [a combined diff](https://fifo.ci/version/9a6138040151068ce3c48610c89a135d53f3c90b/), while here's [a diff for this PR only](https://fifo.ci/version/72f8a1c08f7c04880f016f720419d444d4a99f77/).  (I also fixed a separate issue [affecting 3 fifologs](https://fifo.ci/version/88f81b11546ddd80a1f5c763c28cc87cfcb2d5f3/) involving direct colors).
